File: src/collector.py
"""
ニュース収集モジュール

RSS フィードと Hacker News API から今週のホットトピックスを収集する。
"""

import re
import time
import logging
import requests
import feedparser
from datetime import datetime, timedelta, timezone
from dataclasses import dataclass
from typing import Dict, List, Optional, Sequence

logger = logging.getLogger(__name__)

CATEGORY_LABELS = {
    "tech": "🔬 テクノロジー",
    "japan": "🗾 日本",
    "international": "🌍 国際",
    "science": "🔭 科学",
    "general": "📰 一般",
}


def partition_news_among_personas(
    items: List[NewsItem],
    persona_ids: Sequence[str],
) -> Dict[str, List[NewsItem]]:
    """複数キャスターが同週のネタで被らないよう、ニュースをラウンドロビンで割り当てる。"""
    buckets: Dict[str, List[NewsItem]] = {pid: [] for pid in persona_ids}
    if not persona_ids:
        return buckets
    n = len(persona_ids)
    for i, item in enumerate(items):
        pid = persona_ids[i % n]
        buckets[pid].append(item)
    return buckets


@dataclass
class NewsItem:
    title: str
    description: str
    url: str
    source: str
    published_at: datetime
    category: str = "general"
    language: str = "ja"


class NewsCollector:
    def __init__(self, settings: dict):
        self.news_cfg = settings.get("news", {})
        self.max_topics = settings.get("podcast", {}).get("max_topics", 25)

    def collect(self) -> List[NewsItem]:
        """全ソースからニュースを収集し、重複除去・時系列ソートして返す"""
        items: List[NewsItem] = []
        days_back = self.news_cfg.get("days_back", 7)

        for feed_cfg in self.news_cfg.get("rss_feeds", []):
            try:
                fetched = self._fetch_rss(feed_cfg)
                items.extend(fetched)
                logger.info(f"RSS [{feed_cfg.get('label', feed_cfg['url'])}]: {len(fetched)} 件取得")
            except Exception as e:
                logger.warning(f"RSS 取得失敗 [{feed_cfg.get('label', feed_cfg['url'])}]: {e}")

        if self.news_cfg.get("use_hackernews", True):
            try:
                hn = self._fetch_hackernews()
                items.extend(hn)
                logger.info(f"Hacker News: {len(hn)} 件取得")
            except Exception as e:
                logger.warning(f"Hacker News 取得失敗: {e}")

        cutoff = datetime.now(timezone.utc) - timedelta(days=days_back)
        items = [i for i in items if i.published_at >= cutoff]

        items = self._deduplicate(items)
        items.sort(key=lambda x: x.published_at, reverse=True)

        logger.info(f"合計 {len(items)} 件（上限 {self.max_topics} 件）")
        return items[: self.max_topics]

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _fetch_rss(self, feed_cfg: dict) -> List[NewsItem]:
        feed = feedparser.parse(feed_cfg["url"])
        items = []

        for entry in feed.entries[:15]:
            published_at = self._parse_date(entry)
            description = self._clean_html(
                getattr(entry, "summary", "") or getattr(entry, "description", "")
            )[:400]

            title = entry.get("title", "").strip()
            if not title:
                continue

            items.append(
                NewsItem(
                    title=title,
                    description=description,
                    url=entry.get("link", ""),
                    source=feed.feed.get("title", feed_cfg.get("label", "")),
                    published_at=published_at,
                    category=feed_cfg.get("category", "general"),
                    language=feed_cfg.get("language", "ja"),
                )
            )

        return items

    def _fetch_hackernews(self) -> List[NewsItem]:
        n = self.news_cfg.get("hackernews_stories", 30)
        resp = requests.get(
            "https://hacker-news.firebaseio.com/v0/topstories.json", timeout=10
        )
        story_ids = resp.json()[:n]

        items = []
        for sid in story_ids[:20]:
            try:
                story = requests.get(
                    f"https://hacker-news.firebaseio.com/v0/item/{sid}.json", timeout=5
                ).json()

                if not story or story.get("type") != "story" or not story.get("title"):
                    continue

                published_at = datetime.fromtimestamp(
                    story.get("time", time.time()), tz=timezone.utc
                )
                score = story.get("score", 0)
                comments = story.get("descendants", 0)
                description = story.get("text", "") or f"Score: {score} | Comments: {comments}"
                description = self._clean_html(description)[:400]

                items.append(
                    NewsItem(
                        title=story["title"],
                        description=description,
                        url=story.get("url", f"https://news.ycombinator.com/item?id={sid}"),
                        source="Hacker News",
                        published_at=published_at,
                        category="tech",
                        language="en",
                    )
                )
            except Exception:
                pass

        return items

    def _parse_date(self, entry) -> datetime:
        if hasattr(entry, "published_parsed") and entry.published_parsed:
            try:
                return datetime.fromtimestamp(
                    time.mktime(entry.published_parsed), tz=timezone.utc
                )
            except Exception:
                pass
        if hasattr(entry, "updated_parsed") and entry.updated_parsed:
            try:
                return datetime.fromtimestamp(
                    time.mktime(entry.updated_parsed), tz=timezone.utc
                )
            except Exception:
                pass
        return datetime.now(timezone.utc)

    def _clean_html(self, text: str) -> str:
        text = re.sub(r"<[^>]+>", "", text)
        text = re.sub(r"\s+", " ", text)
        return text.strip()

    def _deduplicate(self, items: List[NewsItem]) -> List[NewsItem]:
        seen: set = set()
        unique = []
        for item in items:
            key = re.sub(r"\s+", "", item.title[:30]).lower()
            if key and key not in seen:
                seen.add(key)
                unique.append(item)
        return unique

File: worker/src/collector.ts
// ニュース収集モジュール — src/collector.py をTypeScriptに移植

import { RSS_FEEDS, FEEDS_BY_CATEGORY, PODCAST_CONFIG, CATEGORY_LABELS, type RssFeed } from "./rss-feeds";
import type { Persona } from "./personas";

export interface NewsItem {
  title: string;
  description: string;
  url: string;
  source: string;
  publishedAt: Date;
  category: string;
  language: string;
}

export async function collectNews(): Promise<NewsItem[]> {
  const items: NewsItem[] = [];
  const cutoff = new Date(Date.now() - PODCAST_CONFIG.daysBack * 24 * 60 * 60 * 1000);

  // RSS フィードを並行取得
  const rssResults = await Promise.allSettled(
    RSS_FEEDS.map((feed) => fetchRss(feed))
  );
  for (const result of rssResults) {
    if (result.status === "fulfilled") {
      items.push(...result.value);
    }
  }

  // Hacker News
  try {
    const hnItems = await fetchHackerNews(PODCAST_CONFIG.hackernewsStories);
    items.push(...hnItems);
  } catch {
    // HN取得失敗は無視
  }

  // 期間フィルタ
  const recent = items.filter((i) => i.publishedAt >= cutoff);

  // 重複除去
  const deduped = deduplicate(recent);

  // 時系列降順
  deduped.sort((a, b) => b.publishedAt.getTime() - a.publishedAt.getTime());

  return deduped.slice(0, PODCAST_CONFIG.maxTopics);
}

/**
 * 複数キャスターが同じ週のネタで被らないよう、ニュースをラウンドロビンで割り当てる。
 * items は時系列降順を想定（先頭ほど新しい）。
 */
export function partitionNewsAmongPersonas(
  items: NewsItem[],
  personaIds: readonly string[]
): Record<string, NewsItem[]> {
  const buckets: Record<string, NewsItem[]> = {};
  for (const id of personaIds) buckets[id] = [];
  if (personaIds.length === 0) return buckets;

  for (let i = 0; i < items.length; i++) {
    const id = personaIds[i % personaIds.length];
    buckets[id].push(items[i]);
  }
  return buckets;
}

// RSSフィードを取得してNewsItemリストを返す
async function fetchRss(feed: RssFeed): Promise<NewsItem[]> {
  const resp = await fetch(feed.url, {
    headers: { "User-Agent": "PodcastBot/1.0" },
    signal: AbortSignal.timeout(8000),
  });
  if (!resp.ok) return [];

  const xml = await resp.text();
  return parseRss(xml, feed);
}

function parseRss(xml: string, feed: RssFeed): NewsItem[] {
  const items: NewsItem[] = [];

  // <item> タグを抽出
  const itemMatches = xml.match(/<item[\s>][\s\S]*?<\/item>/gi) ?? [];

  for (const itemXml of itemMatches.slice(0, 15)) {
    const title = extractTag(itemXml, "title");
    if (!title) continue;

    const description = cleanHtml(
      extractTag(itemXml, "description") ?? extractTag(itemXml, "summary") ?? ""
    ).slice(0, 400);

    const url =
      extractTag(itemXml, "link") ??
      extractAttr(itemXml, "enclosure", "url") ??
      "";

    const pubDate =
      extractTag(itemXml, "pubDate") ??
      extractTag(itemXml, "published") ??
      extractTag(itemXml, "dc:date") ??
      "";

    const publishedAt = pubDate ? new Date(pubDate) : new Date();

    items.push({
      title: title.trim(),
      description,
      url,
      source: feed.label,
      publishedAt: isNaN(publishedAt.getTime()) ? new Date() : publishedAt,
      category: feed.category,
      language: feed.language,
    });
  }

  return items;
}

interface HNStory {
  id: number;
  type: string;
  title: string;
  url?: string;
  text?: string;
  score?: number;
  descendants?: number;
  time?: number;
}

// Hacker News API からトップストーリーを取得
async function fetchHackerNews(n: number): Promise<NewsItem[]> {
  const idsResp = await fetch(
    "https://hacker-news.firebaseio.com/v0/topstories.json",
    { signal: AbortSignal.timeout(8000) }
  );
  const ids: number[] = await idsResp.json();
  const targetIds = ids.slice(0, n);

  const storyResults = await Promise.allSettled(
    targetIds.map((id) =>
      fetch(`https://hacker-news.firebaseio.com/v0/item/${id}.json`, {
        signal: AbortSignal.timeout(5000),
      }).then((r) => r.json() as Promise<HNStory>)
    )
  );

  const items: NewsItem[] = [];
  for (const result of storyResults) {
    if (result.status !== "fulfilled") continue;
    const story = result.value;
    if (!story || story.type !== "story" || !story.title) continue;

    const score = story.score ?? 0;
    const comments = story.descendants ?? 0;
    const description = cleanHtml(story.text ?? `Score: ${score} | Comments: ${comments}`).slice(0, 400);

    items.push({
      title: story.title,
      description,
      url: story.url ?? `https://news.ycombinator.com/item?id=${story.id}`,
      source: "Hacker News",
      publishedAt: new Date((story.time ?? Date.now() / 1000) * 1000),
      category: "tech",
      language: "en",
    });
  }

  return items;
}

// ユーティリティ
function extractTag(xml: string, tag: string): string | null {
  const patterns = [
    new RegExp(`<${tag}[^>]*><!\\[CDATA\\[([\\s\\S]*?)\\]\\]><\\/${tag}>`, "i"),
    new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, "i"),
  ];
  for (const re of patterns) {
    const m = xml.match(re);
    if (m) return m[1].trim();
  }
  return null;
}

function extractAttr(xml: string, tag: string, attr: string): string | null {
  const re = new RegExp(`<${tag}[^>]*${attr}="([^"]*)"`, "i");
  const m = xml.match(re);
  return m ? m[1] : null;
}

function cleanHtml(text: string): string {
  return text
    .replace(/<[^>]+>/g, "")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#039;/g, "'")
    .replace(/\s+/g, " ")
    .trim();
}

function deduplicate(items: NewsItem[]): NewsItem[] {
  const seen = new Set<string>();
  return items.filter((item) => {
    const key = item.title.slice(0, 30).toLowerCase().replace(/\s+/g, "");
    if (!key || seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

interface DevToArticle {
  title: string;
  url: string;
  description: string;
  published_at: string;
}

// Dev.to API からテック記事を取得（APIキー不要）
async function fetchDevToArticles(): Promise<NewsItem[]> {
  try {
    const resp = await fetch("https://dev.to/api/articles?top=7&per_page=20", {
      headers: { "User-Agent": "PodcastBot/1.0" },
      signal: AbortSignal.timeout(8000),
    });
    if (!resp.ok) return [];
    const articles: DevToArticle[] = await resp.json();
    return articles.map((a) => ({
      title: a.title,
      description: (a.description ?? "").slice(0, 400),
      url: a.url,
      source: "Dev.to",
      publishedAt: a.published_at ? new Date(a.published_at) : new Date(),
      category: "tech",
      language: "en",
    }));
  } catch {
    return [];
  }
}

/**
 * キャスターの expertise に合わせた専用ニュースを取得する。
 * 各キャスターが異なるソースから記事を取得するため、放送回ごとのトピック被りを防ぐ。
 */
export async function fetchNewsForPersona(persona: Persona): Promise<NewsItem[]> {
  const cutoff = new Date(Date.now() - PODCAST_CONFIG.daysBack * 24 * 60 * 60 * 1000);
  const items: NewsItem[] = [];

  // expertise に合致するカテゴリのフィードのみ取得
  const targetFeeds: RssFeed[] = persona.expertise.flatMap(
    (cat) => FEEDS_BY_CATEGORY[cat] ?? []
  );

  const rssResults = await Promise.allSettled(
    targetFeeds.map((feed) => fetchRss(feed))
  );
  for (const result of rssResults) {
    if (result.status === "fulfilled") items.push(...result.value);
  }

  // tech カテゴリのキャスターには Hacker News + Dev.to を追加
  if (persona.expertise.includes("tech")) {
    try {
      const hnItems = await fetchHackerNews(PODCAST_CONFIG.hackernewsStories);
      items.push(...hnItems);
    } catch { /* HN取得失敗は無視 */ }

    const devToItems = await fetchDevToArticles();
    items.push(...devToItems);
  }

  const recent = items.filter((i) => i.publishedAt >= cutoff);
  const deduped = deduplicate(recent);
  deduped.sort((a, b) => b.publishedAt.getTime() - a.publishedAt.getTime());
  return deduped.slice(0, PODCAST_CONFIG.maxTopics);
}

export interface BuildNewsPromptOptions {
  /** 真のとき：他キャスターと題材が被らない専用割当である旨を明示 */
  exclusiveAssignment?: boolean;
}

// ニュースをカテゴリ別にまとめてプロンプト用テキストに変換
export function buildNewsPrompt(items: NewsItem[], options?: BuildNewsPromptOptions): string {
  const exclusive = options?.exclusiveAssignment ?? false;

  const byCategory: Record<string, NewsItem[]> = {};
  for (const item of items) {
    (byCategory[item.category] ??= []).push(item);
  }

  const lines: string[] = [];

  if (exclusive) {
    lines.push(
      "【担当割当トピック】同一天に収録される他キャスターと題材が被らないよう、あなた専用に振り分けられたニュースだけをリストしています。",
      "・このリストに含まれるニュースだけを、本編の主要トピックとして取り上げてください（イントロのひと言の比喩として他を触れてもよいが、本題は割当分のみ）。",
      "・リスト外の「今週の大きな出来事」に便乗した重複解説はしないでください。",
      ""
    );
  }

  lines.push(
    "今週のニュースリストです。以下を参考に、今週のエピソードを作成してください。",
    exclusive
      ? "（上記のルールに従い、このリストの中から3〜4件程度を選び、深く語ってください。件数が少ない場合はリスト内のすべてを扱ってください）"
      : "（すべてを取り上げる必要はありません。最も興味深いものを自由に選んでください）"
  );

  for (const [category, catItems] of Object.entries(byCategory)) {
    const label = CATEGORY_LABELS[category] ?? category;
    lines.push("", label, "─".repeat(30));
    for (const item of catItems.slice(0, 8)) {
      const langNote = item.language === "en" ? "（英語記事）" : "";
      lines.push(`・${item.title}${langNote}`);
      if (item.description) lines.push(`  → ${item.description.slice(0, 200)}`);
      lines.push(`  出典: ${item.source}`);
    }
  }

  lines.push("", "", "それでは、エピソードのスクリプトを作成してください。");
  return lines.join("\n");
}

// キャスター専用：得意分野に応じたニュースプロンプトを生成
export function buildNewsPromptForPersona(items: NewsItem[], expertiseCategories: string[]): string {
  const byCategory: Record<string, NewsItem[]> = {};
  for (const item of items) {
    (byCategory[item.category] ??= []).push(item);
  }

  const lines: string[] = [
    "今週のニュースリストです。",
    `あなたの得意分野は【${expertiseCategories.map(c => CATEGORY_LABELS[c] ?? c).join("、")}】です。`,
    "これらの分野を特に重視しながら、以下を参考に今週のエピソードを作成してください。",
    "（すべてを取り上げる必要はありません。得意分野から最も興味深いものを優先的に選んでください）",
  ];

  // 得意分野を先に表示
  for (const category of expertiseCategories) {
    if (byCategory[category]) {
      const label = CATEGORY_LABELS[category] ?? category;
      lines.push("", `★ ${label} （あなたの得意分野）`, "─".repeat(30));
      for (const item of byCategory[category].slice(0, 8)) {
        const langNote = item.language === "en" ? "（英語記事）" : "";
        lines.push(`・${item.title}${langNote}`);
        if (item.description) lines.push(`  → ${item.description.slice(0, 200)}`);
        lines.push(`  出典: ${item.source}`);
      }
    }
  }

  // その他のカテゴリを表示
  for (const [category, catItems] of Object.entries(byCategory)) {
    if (!expertiseCategories.includes(category)) {
      const label = CATEGORY_LABELS[category] ?? category;
      lines.push("", label, "─".repeat(30));
      for (const item of catItems.slice(0, 8)) {
        const langNote = item.language === "en" ? "（英語記事）" : "";
        lines.push(`・${item.title}${langNote}`);
        if (item.description) lines.push(`  → ${item.description.slice(0, 200)}`);
        lines.push(`  出典: ${item.source}`);
      }
    }
  }

  lines.push("", "", "それでは、エピソードのスクリプトを作成してください。");
  return lines.join("\n");
}
